File: TAPScala/src/MapReduce/MapReduce.scala
package MapReduce

import java.io.{BufferedReader, File, FileReader, IOException}
import java.util
import java.util.{HashMap, Map}
import java.util.function.Function

object MapReduce {

  def leer(title: String): String = {
    val file = new java.io.File(title)
    val bufferedReader = new java.io.BufferedReader(new java.io.FileReader(file))
    val content = new StringBuilder()
    var line: String = null
    while ({ line = bufferedReader.readLine(); line != null }) {
      content.append(line).append("\n")
    }
    bufferedReader.close()
    content.toString()
  }

  val countWords: Function[String, Int] = (input: Object) => {
    val stringText: String = input.asInstanceOf[String]
    if (stringText == null || stringText.isEmpty) {
      0
    } else {
      stringText.split("\\s+").length
    }
  }

  var wordCount: Function[AnyRef, AnyRef] = (text: AnyRef) => {
    val inputText: String = text.asInstanceOf[String]
    val wordCounts: util.Map[String, Integer] = new util.HashMap[String, Integer]
    val words: Array[String] = inputText.split("\\s+")
    for (word <- words) {
      if (word.nonEmpty) wordCounts.put(word, wordCounts.getOrDefault(word, 0) + 1)
    }
    wordCounts
  }
}

File: Tap/src/MapReduce/MapReduce.java
package MapReduce;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Function;

public class MapReduce {

    /**
     * Applies a word counting function to a text input and returns a map of words and their corresponding counts.
     */
    public static Function<Object, Object> wordCount = (Object text) -> {
        String inputText = (String) text;
        Map<String, Integer> wordCounts = new HashMap<>();
        String[] words = inputText.split("\\s+");
        for (String word : words) {
            if (!word.isEmpty()) {
                wordCounts.put(word, wordCounts.getOrDefault(word, 0) + 1);
            }
        }
        return wordCounts;
    };





    /**
     * Counts the number of words in a text input.
     */
    public static Function<Object, Object> countWords = (Object input) -> {
        String stringText = (String) input;
        if (stringText == null || stringText.isEmpty()) {
            return 0;
        }
        return stringText.split("\\s+").length;
    };

    /**
     * Reads the contents of a file and returns the entire text as a single string.
     *
     * @param title The title of the file to read
     * @return The contents of the file as a single string
     * @throws IOException If an I/O error occurs while reading the file
     */
    public static String leer(String title) throws IOException {
        File file = new File(title);
        BufferedReader bufferedReader = new BufferedReader(new FileReader(file));
        String line;
        StringBuilder content = new StringBuilder();
        while((line = bufferedReader.readLine()) != null){
            content.append(line).append("\n");
        }
        bufferedReader.close();
        return content.toString();
    }
}
